libcst/codemod/commands/rename.py

```diff
@@ -0,0 +1,192 @@
+        )
+
+        self.new_module: Optional[
+            str
+        ] = new_module if new_module is not None else orig_module
+        self.new_object: str = new_object
+        self.new_name: str = (
+            f"{new_module}.{new_object}" if new_module is not None else new_object
+        )
+
+        self.source: QualifiedNameSource = (
+            QualifiedNameSource.IMPORT
+            if orig_module is not None
+            else QualifiedNameSource.LOCAL
+        )
+
+    def leave_Import(
+        self, original_node: cst.Import, updated_node: cst.Import
+    ) -> Union[cst.Import, cst.RemovalSentinel]:
+        for import_alias in original_node.names:
+            name = get_full_name_for_node(import_alias.name)
+            if name is not None and name == self.orig_module:
+                # Remove old import altogether
+                return cst.RemoveFromParent()
+
+        return updated_node
+
+    def leave_ImportFrom(
+        self, original_node: cst.ImportFrom, updated_node: cst.ImportFrom
+    ) -> Union[cst.ImportFrom, cst.RemovalSentinel]:
+        names = original_node.names
+
+        import_module = original_node.module
+        import_module_name = (
+            get_full_name_for_node(import_module) if import_module is not None else None
+        )
+        if (
+            import_module_name is None
+            or import_module_name != self.orig_module
+            or not isinstance(names, Sequence)
+        ):
+            return updated_node
+
+        if (
+            len(names) == 1
+            and get_full_name_for_node(names[0].name) == self.orig_object
+        ):
+            # Remove this ImportFrom altogether
+            return cst.RemoveFromParent()
+
+        new_names = []
+        for import_alias in names:
+            qualname = (
+                f"{import_module_name}.{get_full_name_for_node(import_alias.name)}"
+            )
+            if qualname != self.orig_name:
+                new_names.append(import_alias)
+
+        return updated_node.with_changes(names=new_names)
+
+    def leave_Attribute(
+        self, original_node: cst.Attribute, updated_node: cst.Attribute
+    ) -> cst.Attribute:
+        if QualifiedNameProvider.has_name(
+            self, original_node, QualifiedName(name=self.orig_name, source=self.source),
+        ):
+            name = get_full_name_for_node(original_node)
+            if name is None:
+                raise Exception("Logic error!")
+            module_name = ".".join(name.split(".")[:-1])
+            module_asname = (
+                None
+                if module_name == self.orig_module
+                else ".".join(name.split(".")[:-1])
+            )
+
+            if self.new_module is not None:
+                AddImportsVisitor.add_needed_import(
+                    self.context,
+                    # pyre-ignore[6]: Expected `str` for 2nd parameter `module` to call
```

Comment:
Instead of pyre-ignore, these can be silenced with something like `module = self.new_module; assert module is not None`

Reply:
Correct. We don't usually use pyre-ignore unless it's something Pyre cannot understand. When adding a pyre-ignore, we usually also report the issue to Pyre team as feature request.
Other than use `assert` to teach Pyre a variable is not None, we usually use an if check since assert is a noop in production mode unlike an `if`.
E.g.
```
module = self.new_module
if module is not None:
    # Pyre won't complain module is an `Optional` in this scope.
```